src/Twig/RolesMatrixExtension.php

```diff
@@ -0,0 +1,99 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of the Sonata Project package.
+ *
+ * (c) Thomas Rabaix <thomas.rabaix@sonata-project.org>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sonata\UserBundle\Twig;
+
+use Sonata\UserBundle\Security\RolesBuilder\MatrixRolesBuilderInterface;
+use Symfony\Component\Form\FormView;
+use Twig\Environment;
+use Twig\Extension\AbstractExtension;
+use Twig\TwigFunction;
+
+/**
+ * @author Christian Gripp <mail@core23.de>
+ * @author Cengizhan Çalışkan <cengizhancaliskan@gmail.com>
+ * @author Silas Joisten <silasjoisten@hotmail.de>
+ */
+final class RolesMatrixExtension extends AbstractExtension
+{
+    /**
+     * @var MatrixRolesBuilderInterface
+     */
+    private $rolesBuilder;
+
+    public function __construct(MatrixRolesBuilderInterface $rolesBuilder)
+    {
+        $this->rolesBuilder = $rolesBuilder;
+    }
+
+    public function getFunctions(): array
+    {
+        return [
+            new TwigFunction('renderMatrix', [$this, 'renderMatrix'], ['needs_environment' => true]),
+            new TwigFunction(
+                'renderRolesList',
+                [$this, 'renderRolesList'],
+                ['needs_environment' => true]
+            ),
+        ];
+    }
+
+    public function getName(): string
+    {
+        return self::class;
+    }
+
+    public function renderRolesList(Environment $environment, FormView $form): ?string
+    {
+        $roles = $this->rolesBuilder->getRoles();
+        foreach ($roles as $role => $attributes) {
+            if (isset($attributes['admin_label'])) {
+                unset($roles[$role]);
+                continue;
+            }
+
+            $roles[$role] = $attributes;
+            foreach ($form->getIterator() as $child) {
+                if ($child->vars['value'] == $role) {
+                    $roles[$role]['form'] = $child;
+                }
+            }
+        }
+
+        return $environment->render('@SonataUser/Form/roles_matrix_list.html.twig', [
+            'roles' => $roles,
+        ]);
+    }
+
+    public function renderMatrix(Environment $environment, FormView $form): ?string
+    {
+        $groupedRoles = [];
+        foreach ($this->rolesBuilder->getRoles() as $role => $attributes) {
+            if (!isset($attributes['admin_label'])) {
+                continue;
+            }
+
+            $groupedRoles[$attributes['admin_label']][$role] = $attributes;
+            foreach ($form->getIterator() as $child) {
+                if ($child->vars['value'] == $role) {
+                    $groupedRoles[$attributes['admin_label']][$role]['form'] = $child;
+                }
+            }
+        }
+
+        return $environment->render('@SonataUser/Form/roles_matrix.html.twig', [
```

Comment:
`render` always returns string https://github.com/twigphp/Twig/blob/2.x/lib/Twig/Environment.php#L281  and your return type here is `?string`, should it be `string`?

Reply:
You are right. My tests where wrong. i fixed it